staging/src/k8s.io/externaljwt/pkg/plugin/keycache.go

```diff
@@ -0,0 +1,215 @@
+			return
+		case <-timer.C:
+		}
+
+		timedCtx, cancel := context.WithTimeout(ctx, keySyncTimeout)
+		if err := p.syncKeys(timedCtx); err != nil {
+			klog.Errorf("when syncing supported public keys(Stale set of keys will be supported): %v", err)
+			timer.Reset(fallbackRefreshDuration)
+		} else {
+			timer.Reset(p.nextRefreshHint.Load().Sub(time.Now()))
+			p.broadcastUpdate()
+		}
+		cancel()
+	}
+}
+
+func (p *keyCache) AddListener(listener externaljwt.Listener) {
+	p.listenersLock.Lock()
+	defer p.listenersLock.Unlock()
+
+	p.listeners = append(p.listeners, listener)
+}
+
+func (p *keyCache) GetCacheAgeMaxSeconds() int {
+	val := int(p.nextRefreshHint.Load().Sub(time.Now()).Seconds())
+	if val < 0 {
+		return 0
+	}
+	return val
+}
+
+// GetPublicKeys returns the public key corresponding to requested keyID.
+// Getter is expected to return All keys for keyID ""
+func (p *keyCache) GetPublicKeys(ctx context.Context, keyID string) []externaljwt.PublicKey {
+	pubKeys, ok := p.findKeyForKeyID(keyID)
+	if ok {
+		return pubKeys
+	}
+
+	// If we didn't find it, trigger a sync.
+	if err := p.syncKeys(ctx); err != nil {
+		klog.ErrorS(err, "Error while syncing keys")
+		return []externaljwt.PublicKey{}
+	}
+
+	pubKeys, ok = p.findKeyForKeyID(keyID)
+	if ok {
+		return pubKeys
+	}
+
+	// If we still didn't find it, then it's an unknown keyID.
+	klog.Errorf("Key id %q not found after refresh", keyID)
+	return []externaljwt.PublicKey{}
+}
+
+func (p *keyCache) findKeyForKeyID(keyID string) ([]externaljwt.PublicKey, bool) {
+	if len(p.publicKeys.Load().Keys) == 0 {
+		klog.Error("No keys currently in cache. Initial fill has not completed")
+		return nil, false
+	}
+
+	if keyID == "" {
+		return p.publicKeys.Load().Keys, true
+	}
+
+	for _, key := range p.publicKeys.Load().Keys {
+		if key.KeyID == keyID {
+			return []externaljwt.PublicKey{key}, true
+		}
+	}
+
+	return nil, false
+}
+
+// sync supported external keys.
+// completely re-writes the set of supported keys.
+func (p *keyCache) syncKeys(ctx context.Context) error {
+	_, err, _ := p.syncGroup.Do("", func() (any, error) {
+		newPublicKeys, refreshHintSec, err := p.getTokenVerificationKeys(ctx)
+		if err != nil {
+			externaljwtmetrics.RecordKeysFetchFailure(err)
+			return nil, fmt.Errorf("while fetching token verification keys: %w", err)
+		}
+		externaljwtmetrics.RecordKeysFetchSuccess()
+
+		// Validate and update the refresh hint.
+		if refreshHintSec < 0 {
+			return nil, fmt.Errorf("found invalid refresh hint (%ds)", refreshHintSec)
+		}
+		newHint := time.Now().Add(time.Duration(refreshHintSec) * time.Second)
+		p.nextRefreshHint.Store(&newHint)
+
+		p.publicKeys.Store(newPublicKeys)
+
+		return newPublicKeys, nil
+	})
+	if err != nil {
+		return fmt.Errorf("in singleflight: %w", err)
+	}
+
+	return nil
+}
+
+func (p *keyCache) broadcastUpdate() {
+	p.listenersLock.Lock()
+	defer p.listenersLock.Unlock()
+
+	for _, l := range p.listeners {
+		l.Enqueue()
```

Comment:
I think this could get stuck in an infinite loop if we aren't careful.  We should only issue updates when the actual keys change.

Reply:
I think it's sufficient to make this conditional on the dataTimestamp from the backend changing

Reply:
this only fires from the periodic sync... I don't think an infinite loop is possible

Reply:
Now we are only publishing if Data_timestamp returned from external signer changes.